merge #17812, #16702, #16730, #18023: supplying and using asmap to improve IP bucketing in addrman

kittywhiskers opened the pull request:
## Overview

As part of https://github.com/dashpay/dash/pull/4025, component pull requests that are required by TorV3 logic are being broken down into branches wherever feasable. This PR deals with asmap related address logic.

## Contents

* Merge https://github.com/bitcoin/bitcoin/pull/16702 ("p2p: supplying and using asmap to improve IP bucketing in addrman" by naumenkogs)
* Merge https://github.com/bitcoin/bitcoin/pull/16730 ("Support serialization of std::vector<bool>" by sipa)
* Merge https://github.com/bitcoin/bitcoin/pull/17812 ("config, net, test: asmap feature refinements and functional tests" by fanquake)

## Disclosures

* Dash-specific changes may not have been tested, only compilation has been ensured. Running the client on (a) testnet(s) may be necessary.
* Contains squashed patches bb8958d, 7d2eb46d97bf4ffe88dc97324d5f27de51c2ddaf and 9ad5d1d from UdjinM6

UdjinM6 commented on src/netaddress.cpp:
```diff
@@ -338,9 +339,17 @@ bool CNetAddr::GetIn6Addr(struct in6_addr* pipv6Addr) const
-// get canonical identifier of an address' group
-// no two connections will be attempted to addresses with the same group
-std::vector<unsigned char> CNetAddr::GetGroup() const
+/**
+ * Get the canonical identifier of our network group
+ *
+ * The groups are assigned in a way where it should be costly for an attacker to
+ * obtain addresses with many different group identifiers, even if it is cheap
+ * to obtain addresses with the same identifier.
+ *
+ * @note No two connections will be attempted to addresses with the same network
+ *       group.
+ */
```
this is from some another PR it seems, no?

kittywhiskers replied:
It's from https://github.com/naumenkogs/bitcoin/commit/303372c41a8d5c58a46cf9ed595e30e67bd0bc99 but to prevent merge conflicts, I just added it in

UdjinM6 commented on src/netaddress.cpp:
```diff
@@ -423,14 +461,14 @@ std::vector<unsigned char> CNetAddr::GetGroup(const std::vector<bool> &asmap) co
     }
 
-    vchRet.push_back(nClass);
+    vchRet.push_back(net_class);
```
this whole thing (lines 450-471) was dropped in the original PR

kittywhiskers replied:
I don't really follow. I think the last `vchRet.push_back(net_class);` needs to be deleted but I'm not sure.

UdjinM6 replied:
Pls see https://github.com/bitcoin/bitcoin/pull/16702/commits/e4658aa8eaf1629dd5af8cf7b9717a8e72028251#diff-1ce9cdfdcb1a8474c626ebd0d6eb261d2a91d4c6ac1bbed8feb57a94c61b5701L474-L496

kittywhiskers replied:
Resolved in 624f474391cac37594326bc6400facd385309b1e

UdjinM6 replied:
Hmm... I think `push_back` should go away too https://github.com/bitcoin/bitcoin/pull/16702/files#diff-1ce9cdfdcb1a8474c626ebd0d6eb261d2a91d4c6ac1bbed8feb57a94c61b5701L473-L474

kittywhiskers replied:
Done in 9a7881116193aeffb11b3f973a68283b28755423

UdjinM6 commented on src/addrman.cpp:
```diff
@@ -13,15 +13,21 @@ int CAddrInfo::GetTriedBucket(const uint256& nKey, const std::vector<bool> &asma
     uint64_t hash1 = (CHashWriter(SER_GETHASH, 0) << nKey << GetKey()).GetHash().GetCheapHash();
     uint64_t hash2 = (CHashWriter(SER_GETHASH, 0) << nKey << GetGroup(asmap) << (hash1 % ADDRMAN_TRIED_BUCKETS_PER_GROUP)).GetHash().GetCheapHash();
-    return hash2 % ADDRMAN_TRIED_BUCKET_COUNT;
+    int tried_bucket = hash2 % ADDRMAN_TRIED_BUCKET_COUNT;
```
missed `#include <logging.h>` and a bunch of include cleanups from the original PR

kittywhiskers replied:
Didn't know Dash had `logging.h` so must have removed it by force of habit. I've intentionally skipped the include cleanups, they should be done as a separate PR imo

PastaPastaPasta replied:
We should do the misc cleanups, there's no reason I can think of to do this in another PR

kittywhiskers replied:
It should be done in another PR as header cleanups repo-wide would involve ensuring alphabetical order (see [init.h](https://github.com/bitcoin/bitcoin/blob/b8d0412b213df18f23bf8677ab94068c6cca9f51/src/init.cpp#L12-L55)) but even then it isn't strictly consistent like [here](https://github.com/bitcoin/bitcoin/blob/b8d0412b213df18f23bf8677ab94068c6cca9f51/src/init.cpp#L56) so I think it merits a scripted-diff PR

PastaPastaPasta replied:
I disagree, there's no reason not to do these specific changes in this PR to remain close to upstream.

UdjinM6 commented on src/init.cpp:
```diff
@@ -555,7 +555,7 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-torcontrol=<ip>:<port>", strprintf(_("Tor control port to use if onion listening enabled (default: %s)"), DEFAULT_TOR_CONTROL));
     strUsage += HelpMessageOpt("-torpassword=<pass>", _("Tor control port password (default: empty)"));
-    strUsage += HelpMessageOpt("-asmap=<file>", _("Specify asn mapping used for bucketing of the peers. Path should be relative to the -datadir path."));
+    strUsage += HelpMessageOpt("-asmap=<file>", strprintf(_("Specify asn mapping used for bucketing of the peers (default: %s). Path should be relative to the -datadir path."), DEFAULT_ASMAP_FILENAME));
```
pls move it above `-banscore` like in the original PR to keep cmd line options sorted properly

kittywhiskers replied:
Resolved in https://github.com/dashpay/dash/pull/4028/commits/6e1a7318a715c61ae1b737ed19e18cad35448f17

UdjinM6 commented on test/functional/feature_asmap.py:
```diff
@@ -0,0 +1,83 @@
+
+Verify node behaviour and debug log when launching bitcoind in these cases:
+
+1. `bitcoind` with no -asmap arg, using /16 prefix for IP bucketing
```
pls Dashify this file (`s/bitcoind/dashd/g`)

UdjinM6 reviewed (changes requested):
Thanks! 👍 

Looks _mostly_ ok, see below

kittywhiskers commented on src/Makefile.am:
```diff
@@ -260,6 +260,7 @@ BITCOIN_CORE_H = \
   utilmemory.h \
   utilmoneystr.h \
   utiltime.h \
+  utilasmap.h \
```
Resolved in https://github.com/dashpay/dash/pull/4028/commits/f0bc73165206cfda2e15b4475a2f61c9485a514b

kittywhiskers commented on src/Makefile.am:
```diff
@@ -580,6 +581,7 @@ libdash_util_a_SOURCES = \
   utilmoneystr.cpp \
   utilstrencodings.cpp \
   utiltime.cpp \
+  utilasmap.cpp \
```
Resolved in https://github.com/dashpay/dash/pull/4028/commits/f0bc73165206cfda2e15b4475a2f61c9485a514b

UdjinM6 commented on test/functional/feature_asmap.py:
```diff
@@ -0,0 +1,106 @@
+
+5. `dashd -asmap` with no file specified and a missing default asmap file
+
+6. `bitcoind -asmap` with an empty (unparsable) default asmap file
```
```suggestion
6. `dashd -asmap` with an empty (unparsable) default asmap file
```

kittywhiskers replied:
The pitfalls on not looking into _all commits_ that modify a file 😿. Resolved in 267426d180f9896f73a4a34c9365835549f11109

UdjinM6 commented on test/functional/feature_asmap.py:
```diff
@@ -0,0 +1,106 @@
+        self.node.assert_start_raises_init_error(extra_args=['-asmap'], expected_msg=msg)
+
+    def test_empty_asmap(self):
+        self.log.info('Test bitcoind -asmap with empty map file')
```
```suggestion
        self.log.info('Test dashd -asmap with empty map file')
```

kittywhiskers replied:
Resolved in 267426d180f9896f73a4a34c9365835549f11109

kittywhiskers commented:
Certain `addrman_tests` are failing as of now

* `error: in "addrman_tests/addrman_tried_collisions": check addrman.size() == i has failed [69 != 73]`
* `error: in "addrman_tests/addrman_evictionworks": check addrman.SelectTriedCollision().ToString() == "[::]:0" has failed`
* `error: in "addrman_tests/addrman_noevict": check addrman.SelectTriedCollision().ToString() == "[::]:0" has failed`

UdjinM6 commented:
Looks like we also need https://github.com/bitcoin/bitcoin/pull/16730 for asmap serialization to work properly.

kittywhiskers commented:
I've already added it at https://github.com/dashpay/dash/pull/4025/commits/cb608ded6f8b99027ab57c327757078a39bc2e1e, it will be merged along with other serialization updates

UdjinM6 commented:
I mean, 16702 simply doesn't work without 16730.

kittywhiskers commented:
It doesn't. I'm currently just working on it. The PRs aren't in order right now.

UdjinM6 reviewed (dismissed):
Seems to be working as expected, slightly tested ACK

xdustinface commented on src/init.cpp:
```diff
@@ -1926,6 +1930,31 @@ bool AppInitMain()
+            InitError(strprintf(_("Could not parse asmap file %s"), asmap_path));
+            return false;
+        }
+        const uint256 asmap_version = SerializeHash(asmap);
+        g_connman->SetAsmap(std::move(asmap));
```
```suggestion
        node.connman->SetAsmap(asmap);
        const uint256 asmap_version = SerializeHash(asmap);
```

Shouldn't this be like ^ ?

kittywhiskers replied:
We don't use `node.connman` be I've swapped it as asked

xdustinface replied:
Right sooo,  what i actually thought it should be while i was opening this suggestion is https://github.com/bitcoin/bitcoin/pull/16702/commits/ec45646de9e62b3d42c85716bfeb06d8f2b507dc#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1829-R1830
but with `g_connman` instead of `node.connman`(my bad) so: 

```
g_connman->SetAsmap(asmap);
const uint256 asmap_version = SerializeHash(asmap);
```

I wanted to have the `move` dropped because it wasn't in the backport PR and i thought its some custom improvement.

And what do you think now how it should look like @UdjinM6?

UdjinM6 replied:
@xdustinface Given the 18023 PR description I think we should backport it anyway. Backporting PRs in the same order as they were merged originally sounds like a better way but I don't mind backporting it later if you feel like this PR has too many backports already.

xdustinface replied:
Ah got it now, im fine adding it here 👍🏻

xdustinface commented on src/netaddress.h:
```diff
@@ -122,8 +122,16 @@ class CNetAddr
         uint64_t GetHash() const;
         bool GetInAddr(struct in_addr* pipv4Addr) const;
-        std::vector<unsigned char> GetGroup() const;
+        std::vector<unsigned char> GetGroup(const std::vector<bool> &asmap) const;
```
```suggestion
```

kittywhiskers replied:
Resolving this nit ended up nearly nuking this branch, thankfully I had a backup somewhere and restored it

xdustinface replied:
You may want to have a look at `git reflog` if you mess something up in your branch's history

xdustinface commented on src/netaddress.h:
```diff
@@ -122,8 +122,16 @@ class CNetAddr
+        // peers in AddrMan bucketing based on the AS infrastructure.
+        // The ip->AS mapping depends on how asmap is constructed.
+        uint32_t GetMappedAS(const std::vector<bool> &asmap) const;
+
```
```suggestion

        std::vector<unsigned char> GetGroup(const std::vector<bool> &asmap) const;
```

This is here in the backport PR. See https://github.com/bitcoin/bitcoin/pull/16702/files#diff-9281caf81406a5c4e11f572f613a4cc9769f39c66579ed75a4c957f30b39d836R88

xdustinface reviewed (changes requested):
Looks good overall 👍 Just two points, see below

PastaPastaPasta commented:
Note that tests are failing

xdustinface commented:
@kittywhiskers The last force push moved `GetGroup` again to the wrong place, no? Was this for some reason or mistake? Also, are you going to add bitcoin#18023 here?

kittywhiskers commented:
@PastaPastaPasta, tests should work now. @xdustinface, I force pushed the last working version because tests started to fail, I've relocated `GetGroup` and added #4016 18023

xdustinface reviewed (dismissed):
utACK, seems to be all fine now 👍 

@kittywhiskers FYI, tests were failing because you swapped the lines like i suggested but you didn't drop the `std::move` which basically invalidated `asmap` then before it was used in `SerializeHash`

```
g_connman->SetAsmap(std::move(asmap));
const uint256 asmap_version = SerializeHash(asmap);
```

UdjinM6 reviewed (dismissed):
👍 

utACK

PastaPastaPasta commented on src/init.cpp:
```diff
@@ -525,7 +525,7 @@ std::string HelpMessage(HelpMessageMode mode)
+    strUsage += HelpMessageOpt("-asmap=<file>", strprintf(_("Specify asn mapping used for bucketing of the peers (default: %s). Relative paths will be prefixed by the net-specific datadir location."), DEFAULT_ASMAP_FILENAME));
     strUsage += HelpMessageOpt("-banscore=<n>", strprintf(_("Threshold for disconnecting misbehaving peers (default: %u)"), DEFAULT_BANSCORE_THRESHOLD));
     strUsage += HelpMessageOpt("-bantime=<n>", strprintf(_("Number of seconds to keep misbehaving peers from reconnecting (default: %u)"), DEFAULT_MISBEHAVING_BANTIME));
     strUsage += HelpMessageOpt("-bind=<addr>", _("Bind to given address and always listen on it. Use [host]:port notation for IPv6"));
```
Seems to be missing https://github.com/bitcoin/bitcoin/commit/81c38a24975f34e5894efe3d1aaf45ff6a8cee4a#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1829-R1838

UdjinM6 replied:
it's here https://github.com/dashpay/dash/pull/4028/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1933-R1947

PastaPastaPasta commented on test/functional/feature_asmap.py:
```diff
@@ -75,7 +75,7 @@ def test_default_asmap(self):
+        msg = "Error: Could not find asmap file '\"{}\"'".format(self.default_asmap)
         self.node.assert_start_raises_init_error(extra_args=['-asmap'], expected_msg=msg)
 
     def run_test(self):
```
missing changes https://github.com/bitcoin/bitcoin/commit/b8d0412b213df18f23bf8677ab94068c6cca9f51#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1836-R1845

kittywhiskers replied:
Already present in https://github.com/dashpay/dash/pull/4028/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1942-R1950

PastaPastaPasta commented on test/functional/feature_asmap.py:
```diff
@@ -75,9 +76,18 @@ def test_default_asmap(self):
         self.log.info('Test dashd -asmap with missing default map file')
         self.stop_node(0)
-        msg = "Error: Could not find asmap file '\"{}\"'".format(self.default_asmap)
+        msg = "Error: Could not find asmap file \"{}\"".format(self.default_asmap)
```
these changes aren't in upstream

PastaPastaPasta commented on src/netaddress.cpp:
```diff
@@ -440,12 +445,15 @@ std::vector<unsigned char> CNetAddr::GetGroup(const std::vector<bool> &asmap) co
+    else {
+        // for the rest of the IPv6 network, use /32 groups
         nBits = 32;
+    }
```
Should fix brackets here (specifically this range, to match upstream commit

kittywhiskers replied:
Headers are eventually normalised [here](https://github.com/dashpay/dash/pull/4028/commits/a343f996c1a36fd0f688a0fb0644f107259e826b#diff-1ce9cdfdcb1a8474c626ebd0d6eb261d2a91d4c6ac1bbed8feb57a94c61b5701L409-R465)

PastaPastaPasta commented on test/functional/test_runner.py:
```diff
@@ -93,7 +93,6 @@
     'rpc_txoutproof.py',
```
9e3770e51522083e150c53b29a1b1cf266d6e8d4 missing changes in init.cpp and test/functional/feature_asmap.py

kittywhiskers replied:
`init.cpp` fixes implemented in https://github.com/dashpay/dash/commit/36c368099a64a1dd888a1207600119e25adba856#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1933-R1957. Commit misattribution due to rebase mixup

PastaPastaPasta commented on src/net.h:
```diff
@@ -475,7 +475,7 @@ friend class CNode;
 
-    void SetAsmap(std::vector<bool> asmap) { addrman.m_asmap = asmap; }
+    void SetAsmap(std::vector<bool> asmap) { addrman.m_asmap = std::move(asmap); }
 
```
7cb3e60d5d5917f0a892271090e1297c7ddd722a seems to be missing a number of changes?

kittywhiskers replied:
Wrong commit mentioned in commit body, the body was for https://github.com/dashpay/dash/pull/4028/commits/9be6d8a887ace6575932e2473bd04d5b7dbf9e4f. https://github.com/bitcoin/bitcoin/commit/d58bcdc4b569a667b6974c3547b7ff6f665afce9's `init.cpp` fix is implemented in https://github.com/dashpay/dash/commit/36c368099a64a1dd888a1207600119e25adba856#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1933-R1957

PastaPastaPasta replied:
So what commit is 74f271963f86fef885bb9a081b6f3d346b9865b7 actually backporting? I'm confused now

kittywhiskers replied:
It's supposed to be https://github.com/bitcoin/bitcoin/commit/d58bcdc4b569a667b6974c3547b7ff6f665afce9

PastaPastaPasta reviewed (changes requested):
See comments

kittywhiskers commented on src/Makefile.test.include:
```diff
@@ -22,7 +22,8 @@ JSON_TEST_FILES = \
+  test/data/asmap.raw
 
 GENERATED_TEST_FILES = $(JSON_TEST_FILES:.json=.json.h) $(RAW_TEST_FILES:.raw=.raw.h)
 
```
Logic for generating headers from raw files already exist 
https://github.com/dashpay/dash/blob/36c368099a64a1dd888a1207600119e25adba856/src/Makefile.test.include#L198-L206

Furthermore, Dash current encapsulates all generated header output into `namespace raw_tests` while the commit uses the `_raw` suffix, changing  No reason to add something that is already implemented.

PastaPastaPasta replied:
Okay, I think that's good

kittywhiskers commented on src/addrman.h:
```diff
@@ -187,7 +193,8 @@ class CAddrInfo : public CAddress
 {
-private:
+friend class CAddrManTest;
+protected:
```
Don't remember, AFAIK it was probably to resolve a compile error

UdjinM6 replied:
Good catch! This should be https://github.com/bitcoin/bitcoin/pull/13115/files#diff-164bd9e2e30f54d0a79eb7cc372309e2f2155edc6c3f051290ab078f03f6a771R190-R194 (we are missing github.com/bitcoin/bitcoin/pull/13115 here which was merged into btc 0.18).

kittywhiskers replied:
Fix in latest push

PastaPastaPasta replied:
I don't believe this actually wa fixed

kittywhiskers replied:
I might've just placed everything that was protected into private, then fixed that. I'll add #13155 to the PR the next morning

kittywhiskers commented on src/init.cpp:
```diff
@@ -553,6 +556,7 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-timeout=<n>", strprintf(_("Specify connection timeout in milliseconds (minimum: 1, default: %d)"), DEFAULT_CONNECT_TIMEOUT));
     strUsage += HelpMessageOpt("-torcontrol=<ip>:<port>", strprintf(_("Tor control port to use if onion listening enabled (default: %s)"), DEFAULT_TOR_CONTROL));
     strUsage += HelpMessageOpt("-torpassword=<pass>", _("Tor control port password (default: empty)"));
+
```
Resolved

kittywhiskers commented on src/init.cpp:
```diff
@@ -1926,6 +1930,31 @@ bool AppInitMain()
 
+    // Read asmap file if configured
+    if (gArgs.IsArgSet("-asmap")) {
+        fs::path asmap_path = fs::path(gArgs.GetArg("-asmap", ""));
+        if (asmap_path.empty()) {
+            asmap_path = DEFAULT_ASMAP_FILENAME;
+        }
+        if (!asmap_path.is_absolute()) {
+            asmap_path = GetDataDir() / asmap_path;
+        }
+        if (!fs::exists(asmap_path)) {
+            InitError(strprintf(_("Could not find asmap file %s"), asmap_path));
+            return false;
+        }
+        std::vector<bool> asmap = CAddrMan::DecodeAsmap(asmap_path);
+        if (asmap.size() == 0) {
+            InitError(strprintf(_("Could not parse asmap file %s"), asmap_path));
+            return false;
+        }
+        const uint256 asmap_version = SerializeHash(asmap);
+        g_connman->SetAsmap(std::move(asmap));
+        LogPrintf("Using asmap version %s for IP bucketing\n", asmap_version.ToString());
```
A blend of various init-specific changes being rebased and eventually mixed into the initial commit, leaving the commits that actually introduce the changes blank.

kittywhiskers commented on src/init.cpp:
```diff
@@ -136,6 +137,7 @@ static CDSNotificationInterface* pdsNotificationInterface = nullptr;
 #endif
 
 static const char* FEE_ESTIMATES_FILENAME="fee_estimates.dat";
+static const char* DEFAULT_ASMAP_FILENAME="ip_asn.map";
```
Resolved in latest push

kittywhiskers commented on src/init.cpp:
```diff
@@ -44,6 +44,7 @@
 #include <util.h>
 #include <utilmoneystr.h>
+#include <utilasmap.h>
 #include <validationinterface.h>
```
Bias towards adding as few headers as possible when resolving merge conflicts

PastaPastaPasta replied:
imo you should include hash.h as well as the newlines that upstream added

kittywhiskers replied:
Newlines have been resolved, I'll add `hash.h`

PastaPastaPasta reviewed (changes requested):
Looks very close to being good!!

16730 utACK
16702 looks good besides maybe https://github.com/dashpay/dash/pull/4028/files/688985b2f4ab78e32e1dfb96640a24e7435a66c1..02976ff105730386804a18f5253fdb8161677be9#r590519900 https://github.com/dashpay/dash/pull/4028/files/688985b2f4ab78e32e1dfb96640a24e7435a66c1..02976ff105730386804a18f5253fdb8161677be9#r596435581


17812 seems to be missing changes in init.cpp https://github.com/bitcoin/bitcoin/pull/17812/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69d

18023 missing https://github.com/bitcoin/bitcoin/pull/18023/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69d

kittywhiskers commented:
> 17812 seems to be missing changes in init.cpp https://github.com/bitcoin/bitcoin/pull/17812/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69d

Already implemented in https://github.com/dashpay/dash/pull/4028/commits/ad5c4a21824f604ef78dfd6421f0e26439d51079#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1934-R1958 (ad5c4a21824f604ef78dfd6421f0e26439d51079)

> 18023 missing https://github.com/bitcoin/bitcoin/pull/18023/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69d

Already implemented in https://github.com/dashpay/dash/pull/4028/commits/ad5c4a21824f604ef78dfd6421f0e26439d51079#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1953 (ad5c4a21824f604ef78dfd6421f0e26439d51079)

---

Addition of `hash.h` to `init.cpp` and moving `m_asmap` below `m_added_nodes` in `net.h` complete.

PastaPastaPasta commented:
> Already implemented in ad5c4a2#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1934-R1958 (ad5c4a2)

Please move these changes from the 18023 commit to a 17812 commit

kittywhiskers commented:
Resolved conflicts, fixed misattribution and squashed commits on a new branch (`asmap_squashed`). Please review. https://github.com/kittywhiskers/dash/tree/asmap_squashed

kittywhiskers commented:
The `asmap` branch should now be fit for review

UdjinM6 commented:
I think you squashed a lot of changes that belong to 16702 and 17812 into 18023 backport instead...

kittywhiskers commented:
> The asmap branch should now be fit for review

This didn't age well... sigh, I'll give it a look

PastaPastaPasta commented on src/init.cpp:
```diff
@@ -511,6 +511,7 @@ void SetupServerArgs()
     gArgs.AddArg("-timestampindex", strprintf("Maintain a timestamp index for block hashes, used to query blocks hashes by a range of timestamps (default: %u)", DEFAULT_TIMESTAMPINDEX), false, OptionsCategory::INDEXING);
```
https://github.com/bitcoin/bitcoin/pull/17812/files moved the `#include <util/asmap.h>` I assume that it was placed in the correct location previously. If thats the case, feel free to resolve this comment, otherwise, please move that include

kittywhiskers replied:
The includes are done in alphabetical order and since includes are their own directory, `asmap.h` comes first but since Dash hasn't backported those changes, the alphabetical order dictates starting with `util.h`, then `utilasmap.h`. The original PR presumes `asmap.h` is located within `util` but such distinction doesn't exist as of now.

PastaPastaPasta commented on src/netaddress.cpp:
```diff
@@ -200,7 +200,21 @@ bool CNetAddr::IsRFC7343() const
+    return (GetByte(15) == 0x20 && GetByte(14) == 0x01 && GetByte(13) == 0x04 && GetByte(12) == 0x70);
+}
+
+/**
+ * @returns Whether or not this is a dummy address that maps an onion address
+ *          into IPv6.
+ *
+ * @see CNetAddr::SetSpecial(const std::string &)
+ */
+bool CNetAddr::IsTor() const
+{
+    return (memcmp(ip, pchOnionCat, sizeof(pchOnionCat)) == 0);
+}
```
These changes are from somewhere different

kittywhiskers replied:
Must me during merge resolution, will look into it

kittywhiskers replied:
Dash's `IsTor()` comes from https://github.com/dashpay/dash/commit/888f36abe7eded6bd78f614fa2b52f5bbfe1c066, which was done way after these commits. Makes sense why there's a conflict.

PastaPastaPasta commented on src/netaddress.h:
```diff
@@ -109,8 +109,9 @@ class CNetAddr
         bool IsRFC4862() const; // IPv6 autoconfig (FE80::/64)
-        bool IsRFC6052() const; // IPv6 well-known prefix (64:FF9B::/96)
-        bool IsRFC6145() const; // IPv6 IPv4-translated address (::FFFF:0:0:0/96)
+        bool IsRFC6052() const; // IPv6 well-known prefix for IPv4-embedded address (64:FF9B::/96)
+        bool IsRFC6145() const; // IPv6 IPv4-translated address (::FFFF:0:0:0/96) (actually defined in RFC2765)
+        bool IsHeNet() const;   // IPv6 Hurricane Electric - https://he.net (2001:0470::/36)
```
This adds more than upsteam did https://github.com/bitcoin/bitcoin/pull/17812/files only adds IsHeNet

kittywhiskers replied:
Aside from modifying comments surrounding `IsRFC6052()` and `IsRFC6145()`, it only adds `IsHeNet()`, would you like those modified comments removed?

PastaPastaPasta replied:
True true :+1: 

Modified comments are fine I guess, although try to avoid that in future

PastaPastaPasta commented on test/functional/feature_asmap.py:
```diff
@@ -0,0 +1,106 @@
+        self.num_nodes = 1
+
+    def test_without_asmap_arg(self):
+        self.log.info('Test bitcoind with no -asmap arg passed')
```
dashify

PastaPastaPasta commented:
You have a linter failure

UdjinM6 commented:
Pls see bb8958d129c90262030683d8fae9f5ac403627de and 9ad5d1d3b90e84563822f7bd7d4ca604df9fa02c

kittywhiskers commented:
@UdjinM6, integrated in 067aa67a9d

UdjinM6 commented on src/Makefile.test.include:
```diff
@@ -205,7 +206,7 @@ test/data/%.raw.h: test/data/%.raw
 	@{ \
 	 echo "namespace raw_tests{" && \
-	 echo "static unsigned const char $(*F)[] = {" && \
+   echo "static unsigned const char $(*F)[] = {" && \
```
Pls revert this one too https://github.com/UdjinM6/dash/commit/7d2eb46d97bf4ffe88dc97324d5f27de51c2ddaf (my bad 🙈 )

kittywhiskers replied:
Resolved in 4b2b5f7

UdjinM6 reviewed (approved):
LGTM, utACK

PastaPastaPasta reviewed (approved):
utACK
